### PR TITLE
[5.6] Logging, as it should be

### PR DESCRIPTION
**Pull request by phroggyy**
This PR rewrites *all* of Laravel's existing logging system. With the new code, users are given the ability to configure their logging setup, and gives much greater flexibility when it comes to third party integration.

*Features:*

- channels: you can have multiple log "channels", allowing you to send certain messages to certain channels. This lets you separate logging into multiple concerns (i.e request logging can go into a `requests` channel, which writes to a daily log file, while your default log writes to `laravel.log`)
- pipes: you can specify arbitrary pipes through which the channel will pass during configuration. By default, there's a pipe that sets the minimum log level on attached handlers; i.e log only error or greater to the `errors` channel, log only `emergency` or greater to the `bedroom-lights` channel
- event logging: `Log::event('some.event')->to('events')` will send a log record to your `events` channel whenever the `some.event` event occurs. The logger will attach all event data it can find (yes, even on class events, it will grab properties)

With this merged, I would PR laravel/laravel to add the following to users' default config:

`config/logging.php`
```
return [
    /*
    |--------------------------------------------------------------------------
    | Default Channels
    |--------------------------------------------------------------------------
    | This option controls the default logging channels to which Laravel will
    | send log messages. You may specify a single channel, or use an array
    | to register multiple default channels if the application needs it.
    |
    */
    'default' => 'laravel',

    /*
    |--------------------------------------------------------------------------
    | Channel Definitions
    |--------------------------------------------------------------------------
    | This option lets you specify all channel definitions, which will be used
    | by Laravel to construct logging channels. A default has been provided
    | for you, which writes all log messages to storage/logs/laravel.log.
    |
    | Every channel must specify a type, which tells laravel what channel type
    | should be used to handle records. You may create custom channel types
    | to customize the log handler, or use one of the available defaults.
    |
    | Available defaults: single, daily, syslog, errorlog
    |
    */
    'channels' => [
        'laravel' => [
            'type' => 'single',
            'file' => storage_path('/logs/laravel.log'),
            'pipes' => 'level:debug',    //for demonstration purposes, would not be included in default config
        ],
        'bedroom-lights' => [    // sample config, would not be included as default
            'type' => PhilipsHueChannel::class,
            'host' => env('HUE_HOST', 'http://localhost:1234'),
            'pipes' => 'fingers-crossed:error'
        ]
    ],
];
```

Pipes can be referenced either by their class name (`FingersCrossed::class`), or by an alias. Currently, aliases are defined in the `LogServiceProvider`, which makes it a bit difficult to register them under custom names (just as difficult as it is to modify the SP in the current state). An improvement on this would be to add a method to register pipes through a separate class, but I left that out here due to complexity.

### CODE REVIEW

**Comment by deleugpn** [comment]
Do you have any plans to keep PSR-3 compliance?

**Comment by phroggyy** [comment]
sure @deleugpn, it is really PSR-3 compliant, as it all proxies down to Monolog. But yes, we could explicitly implement the `Logger` interface on the `LogManager`, and make each channel implement the `LoggerAwareInterface`

**Comment by kamui545** [comment]
Yay! Maybe I will be able to trash my hacky multi-channel logging implementation in the future 👍

File: src/Illuminate/Contracts/Logging/ChannelResolutionException.php [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace Illuminate\Contracts\Logging;
+
+use Exception;
+
+class ChannelResolutionException extends Exception
```
Comment (jmarcher) [inline src/Illuminate/Contracts/Logging/ChannelResolutionException.php]: Why are you defining concrete classes inside the Contracts namespace?

File: src/Illuminate/Contracts/Logging/ChannelUndefinedException.php [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace Illuminate\Contracts\Logging;
+
+use Exception;
+
+class ChannelUndefinedException extends Exception
```
Comment (jmarcher) [inline src/Illuminate/Contracts/Logging/ChannelUndefinedException.php]: Why are you defining concrete classes inside the Contracts namespace?

File: src/Illuminate/Log/LogServiceProvider.php [hunk 3]
```diff
@@ -2,11 +2,30 @@
 
 namespace Illuminate\Log;
 
+use Illuminate\Contracts\Events\Dispatcher;
+use Illuminate\Contracts\Logging\ChannelUndefinedException;
```
Comment (jmarcher) [inline src/Illuminate/Log/LogServiceProvider.php]: uses need to be sorted by length

**Review by jmarcher — CHANGES_REQUESTED** [review]

File: src/Illuminate/Contracts/Logging/ChannelResolutionException.php (hunk 1 above)
Comment (phroggyy) [reply src/Illuminate/Contracts/Logging/ChannelResolutionException.php]: Sorry, that's a mistake, thanks

**Comment by rizqidjamaluddin** [comment]
Should we have a property on class events to specify a default log message for it?

**Comment by phroggyy** [comment]
that's a nice idea @rizqidjamaluddin!

**Comment by gmsantos** [comment]
@phroggyy about the middleware part, isn't the monolog Handler responsible to define witch level of logging it will accept or not? E.g.: 
https://github.com/Seldaek/monolog/blob/master/src/Monolog/Handler/StreamHandler.php#L36
https://github.com/Seldaek/monolog/blob/master/src/Monolog/Handler/AbstractHandler.php#L32

**Comment by phroggyy** [comment]
You're right @gmsantos. I still want to keep the middleware functionality, and I do want to give the users an easy way to configure log level, I guess each channel could just have a `level` setting?

**Comment by rizqidjamaluddin** [comment]
I just feel it's a little more Laravel-y to have the log filtering be in the middleware, and use the `FilterHandler` instead. Keeps all the logic within a log channel to the `middleware` config key.

**Comment by phroggyy** [comment]
I think, for now, I'm gonna go ahead and fix up the current implementation, and additional changes can be submitted as further PRs

File: src/Illuminate/Foundation/Application.php [hunk 4]
```diff
@@ -1019,39 +1012,6 @@ public function provideFacades($namespace)
     }
 
     /**
-     * Define a callback to be used to configure Monolog.
-     *
-     * @param  callable  $callback
-     * @return $this
-     */
-    public function configureMonologUsing(callable $callback)
```
Comment (gmsantos) [inline src/Illuminate/Foundation/Application.php]: I suggest keeping the `configureMonologUsing`, it makes possible to use totally customized log Handlers, Formatters and Channels independently from the framework.

I use it in my case to configure how I gonna handle the app logs, sending to external services via Handlers based on log level, passing through [FingersCrossedHandler](https://github.com/Seldaek/monolog/blob/master/src/Monolog/Handler/FingersCrossedHandler.php) and [WhatFailureGroupHandler](https://github.com/Seldaek/monolog/blob/master/src/Monolog/Handler/WhatFailureGroupHandler.php) and so on.

It's a so specific strategy that I doubt that will be shipped my default on Laravel. Removing this functionality is a :-1: for me.

**Review by gmsantos — CHANGES_REQUESTED** [review]

**Comment by gmsantos** [comment]
> You're right @gmsantos. I still want to keep the middleware functionality, and I do want to give the users an easy way to configure log level, I guess each channel could just have a level setting?

I personally use handlers to define what kind of log goes to each place. [FingersCrossedHandler]( https://github.com/Seldaek/monolog/blob/master/src/Monolog/Handler/FingersCrossedHandler.php) is something really nice to use on a errors channel to debug production logs and a middleware to filter logs could make it useless. 

I like how [MonologBundle](http://symfony.com/doc/3.4/logging.html) from Symfony makes logs simply customizable. Some ideias to build my custom Log config comes from there.

Though I like the idea to use different channels for requests or events (I gonna improve my application log with that :smile:)

File: src/Illuminate/Foundation/Application.php (hunk 4 above)
Comment (phroggyy) [reply src/Illuminate/Foundation/Application.php]: Well you can easily just create your own channel that attaches the handlers to your default logging? That'd be the new way to do it, a new `SuperSpecialChannel` which extends `Channel` and pushes whichever handlers you want onto the monolog instance. The entire reason for this PR is in fact because of how limited `configureMonologUsing` is

**Comment by phroggyy** [comment]
@gmsantos the difference is I want to strike a balance between customisability (which is quite shocking in Laravel at the moment), and usability. Middleware is something users are already used to; to use the `FingersCrossedHandler`, you'd do something like this:

```
class UrgentChannel extends Channel
{
    public function prepare(array $options = [])
    {
        $this->writer->pushHandler(new FingersCrossedHandler(
            new SlackHandler($this->app['config']->get('services.slack.secret'))
        ));
    }
}
```

And in your log config you could do
```
[
    'default' => ['single', 'home-alert']
    'channels' => [
        'home-alert' => [
            'type' => UrgentHandler::class
        ],
        'single' => [
            'type' => 'single'
        ]
    ]
]
```

This would send all your logs to the urgent handler, and let you use fingers crossed as you do today. The benefit of channels is, aside from configurability, the fact that it gets its own monolog instance, which means each channel can have different processors (as processors are attached to the logger instance), which can be beneficial if you want to log loads of additional data for only certain channels.

Again, building your own custom channel gives you full access to the monolog instance so you can do whatever you want! It just ends up in a class instead of `bootstrap/app.php`.

**Review by gmsantos — APPROVED** [review]

**Comment by gmsantos** [comment]
Nice approach with channels @phroggyy. Reviewing full PR an the example config I get your point.

**Comment by phroggyy** [comment]
Glad you like it @gmsantos! Sorry, I should've explained the channels concept better before! If I get this in, which I surely hope, I'll be PRing laravel/laravel too to add the config in as well, and the docs

File: src/Illuminate/Log/LogManager.php [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        return $logger->to($channels);
+    }
+
+    /**
+     * Resolve a channel by name.
+     *
+     * @param  string  $channel
+     * @return \Illuminate\Log\Channel
+     *
+     * @throws \Exception
```
Comment (GrahamCampbell) [inline src/Illuminate/Log/LogManager.php]: Please make this doc more specific.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Log/LogManager.php [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    public function __construct(Dispatcher $events)
+    {
+        $this->events = $events;
+    }
+
+    /**
+     * Retrieve a channel to use for logging.
+     *
+     * @param  string  $channel
+     * @return \Illuminate\Log\Channel
```
Comment (GrahamCampbell) [inline src/Illuminate/Log/LogManager.php]: Surely this should be the channel contract?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Log/LogManager.php (hunk 5 above)
Comment (phroggyy) [reply src/Illuminate/Log/LogManager.php]: Good catch! Thanks, updated.

File: src/Illuminate/Log/LogManager.php (hunk 6 above)
Comment (phroggyy) [reply src/Illuminate/Log/LogManager.php]: Definitely, thanks! Updated

Comment (phroggyy) [reply src/Illuminate/Log/LogManager.php]: Updated all references to the concrete

File: src/Illuminate/Log/LogManager.php [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    /**
+     * Register a new log channel.
+     *
+     * @param  string  $name
+     * @param  \Illuminate\Log\Channel  $channel
+     * @return void
+     */
+    public function registerChannel($name, Channel $channel)
```
Comment (GrahamCampbell) [inline src/Illuminate/Log/LogManager.php]: Shouldn't this be updated to take a chanel contract?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Log/LogManager.php (hunk 7 above)
Comment (phroggyy) [reply src/Illuminate/Log/LogManager.php]: Thanks Graham, good catch

**Comment by taylorotwell** [comment]
I think overall I'm on board with the concepts if you want to clean it all up for final presentation.

**Comment by rizqidjamaluddin** [comment]
Future idea: it'd be really nice if we can get framework-level logging, keeping track of things like app booting, SP registrations, route handling, HTTP/console activity etc. It'd improve debugging anything framework related, identifying package conflicts, configuration issues, environment issues, 3rd party driver activity etc.

Since this is a pretty specific use case, we can have it disabled or ignored by default. It could go into its own default `framework` channel to keep it separate.

**Comment by phroggyy** [comment]
@taylorotwell after discussing with @rizqidjamaluddin, I think I'll make the `middleware` be an option for configuration when the channels are constructed, rather than per-record, since per-record can be done through Monolog processors and handlers, meaning a middleware on construction would let us have something like a `fingers-crossed` middleware that wraps the existing handler(s) of a channel.

I'm not sure I like the name `middleware` though, got any other suggestions?

cc @gmsantos I think this will suit your desire for attaching stuff like fingers-crossed better

**Comment by phroggyy** [comment]
@taylorotwell I changed the behaviour of middleware now to the above. Think it's ready for a final review now. I'll of course gladly take comments.

Another slight change in behaviour was to denormalize config for consistency with other config files in Laravel (custom config is now not under an `extra` key, but rather under the main channel config)

File: src/Illuminate/Log/Channel.php [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        $this->writer->log($level, $message, $context);
+    }
+
+    /**
+     * Pass the channel through middleware for configuration.
+     *
+     * @param  array  $middleware
+     * @return \Illuminate\Contracts\Log\Channel
+     */
+    public function through(array $middleware)
```
Comment (gmsantos) [inline src/Illuminate/Log/Channel.php]: ideia: maybe rename middleware with filter to avoid confusion with http middlewares

File: src/Illuminate/Log/SingleChannel.php [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+     * Setup a new handler for a single file.
+     *
+     * @param  array  $options
+     * @return void
+     */
+    public function prepare(array $options = [])
+    {
+        $options = array_merge($this->defaultOptions, $options);
+
+        $this->writer->pushHandler(new StreamHandler($this->app->storagePath().'/logs/'.$options['file']));
```
Comment (gmsantos) [inline src/Illuminate/Log/SingleChannel.php]: In production we leave our logs files outside Laravel storage path. Our deployment swaps the whole application folder to a symlink, so make it possible to use any path is better to hardcoded storage path.

File: src/Illuminate/Log/Channel.php (hunk 8 above)
Comment (phroggyy) [reply src/Illuminate/Log/Channel.php]: I'm not sure I like filter either; since it now runs *at setup*, it'd make more sense with something like `configuration`, but that doesn't sound good either

File: src/Illuminate/Log/SingleChannel.php (hunk 9 above)
Comment (phroggyy) [reply src/Illuminate/Log/SingleChannel.php]: good shout

File: src/Illuminate/Log/Channel.php (hunk 8 above)
Comment (DaneEveritt) [reply src/Illuminate/Log/Channel.php]: `pipes` perhaps? Thats how `Pipeline` references the variable on `through`.

Comment (phroggyy) [reply src/Illuminate/Log/Channel.php]: Updated this to `pipes` now

**Comment by phroggyy** [comment]
My thought is also that after this is in, I'll do a second PR to add a method on the `Application` that lets you specify which SP should be used to bootstrap logging. Upgrade guide will then tell people to

1. Create a new class `App\Providers\LogServiceProvider`:
```
<?php

use Illuminate\Logging\LogServiceProvider as ServiceProvider;

class LogServiceProvider extends ServiceProvider
{
    //
}
```
2. Add `$app->configureLoggingWith(App\Providers\LogServiceProvider::class)` to `bootstrap/app.php`
3. Add `config/logging.php`

Also, really need a better name for `pipes`... Like `setup` or something

**Comment by taylorotwell** [comment]
Please update your original comment with an example of all new features, including middleware.

**Comment by phroggyy** [comment]
@taylorotwell updated the sample config now to add another channel. That `fingers-crossed` pipe would be added by i.e some package or by the user

**Comment by gmsantos** [comment]
@phroggyy  just mall fix from sample config, in other configs file paths are inside the helper

`'file' => storage_path('logs/laravel.log')',`

**Comment by phroggyy** [comment]
@gmsantos thanks! fixed

**Comment by m1guelpf** [comment]
@phroggyy I know this might be better feedback for the laravel/laravel PR, but why `config/logging.php` and not `config/log.php`? The second seems clearer to me...

**Comment by taylorotwell** [comment]
I guess I have a hard time seeing the actual use case or "pipes"?

**Comment by phroggyy** [comment]
@taylorotwell they're essentially decorators. It means you don't have to create a custom channel when you i.e want to change the log level, or you could have a pipe (again, really dislike the name) that wraps your handlers in a `FingersCrossedHandler`, or you could have a pipe that adds additional processors to your channel's logger.

It's just a way to make it easier to have add-ons, as opposed to needing to create custom channels. i.e this package https://github.com/decahedronio/monolog-sticky-context is just a Monolog processor. We could easily turn this into a pipe so you could do
```
'pipes' => 'sticky'
```
as opposed to having to create a new channel that `extends SingleChannel` or whatever you use, to simply push a processor.

**Comment by iwex** [comment]
@phroggyy any updates?

**Comment by phroggyy** [comment]
@iwex I'm not the person to ask, you'd have to ask Taylor for that.

**Comment by iwex** [comment]
@GrahamCampbell @taylorotwell don't forget about this :)

**Comment by jmarcher** [comment]
@iwex 5.6 still has a long way to go, I do not see the rush.

**Comment by iwex** [comment]
@jmarcher Don't put off until tomorrow what you can do today :)

**Comment by clarkeash** [comment]
This would be great 😄

**Comment by taylorotwell** [comment]
How does someone actually define and register a pipe? Can that be added to the original comment?

**Comment by phroggyy** [comment]
Updated @taylorotwell. In short, users will probably want to just use the class name with the current implementation. That said I'm open to suggestions on better ways to handle it (defining through config perhaps? Not sure if that's clean enough)

File: src/Illuminate/Foundation/Application.php (hunk 4 above)
Comment (JulienTant) [reply src/Illuminate/Foundation/Application.php]: This has to be kept, eventually you can mark it as deprecated but we can't just remove a public method.

**Comment by taylorotwell** [comment]
Thanks. Can you summarize the total amount of breaking changes in this PR?

**Comment by taylorotwell** [comment]
@phroggyy ping. I like the idea behind this PR but will need a few days to clone it down and really work through all of the code very closely. Are you able to summarize all the breaking changes?

File: src/Illuminate/Contracts/Log/Channel.php [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+     * Setup an event listener for automatic logging.
+     *
+     * @param  \Illuminate\Contracts\Events\Dispatcher  $events
+     * @param  string  $event
+     * @param  string  $level
+     * @return void
+     */
+    public function listen(Dispatcher $events, $event, $level = 'debug');
```
Comment (mfn) [inline src/Illuminate/Contracts/Log/Channel.php]: Shouldn't this rather directly reference `\Psr\Log\LogLevel::DEBUG`?

File: src/Illuminate/Contracts/Log/Channel.php [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+     * @return void
+     */
+    public function setFormatter(FormatterInterface $formatter);
+
+    /**
+     * Setup an event listener for automatic logging.
+     *
+     * @param  \Illuminate\Contracts\Events\Dispatcher  $events
+     * @param  string  $event
+     * @param  string  $level
```
Comment (mfn) [inline src/Illuminate/Contracts/Log/Channel.php]: See my comment below, could reference`\Psr\Log\LogLevel`

File: src/Illuminate/Log/DailyChannel.php [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+class DailyChannel extends Channel
+{
+    /**
+     * The default options for the channel.
+     *
+     * @var array
+     */
+    protected $defaultOptions = [
+        'file' => 'laravel.log',
+        'days' => 5,
```
Comment (mfn) [inline src/Illuminate/Log/DailyChannel.php]: Any specific reason for 5? I suggest 7, i.e. a week.

**Comment by mfn** [comment]
@phroggyy Can you given an example how I can register a new formatter?

I only saw the fixed list in `LogServiceProvider::$formats`.

File: src/Illuminate/Log/DailyChannel.php (hunk 12 above)
Comment (phroggyy) [reply src/Illuminate/Log/DailyChannel.php]: This is the default currently used by Laravel

File: src/Illuminate/Contracts/Log/Channel.php (hunk 11 above)
Comment (phroggyy) [reply src/Illuminate/Contracts/Log/Channel.php]: Nope. It'd still be a string. `LogLevel::DEBUG` is a string

File: src/Illuminate/Contracts/Log/Channel.php (hunk 10 above)
Comment (phroggyy) [reply src/Illuminate/Contracts/Log/Channel.php]: We could, I'm not sure it matters a lot, and just having the string may be more clear, not sure.

**Comment by phroggyy** [comment]
@taylorotwell sorry, been busy lately. Breaking changes:

- `configureMonologUsing` is gone
- the custom `Log` contract is gone
- the log `Writer` is gone
- all configuration is now taken from `config/logging.php` rather than `config/app.php`

So basically, everything is gone. From an end-user perspective, it should work as before if you were never binding anything custom to the log contract, or if you did not use `configureMonologUsing`, or any weird trick to customise logging beyond what Laravel made possible.

For those using `configureMonologUsing`, the most sane option will likely be to create a custom channel, or just use an existing channel with custom pipes, depending on the complexity.

For those that used a service provider for more custom configuration like `Log::getMonolog()->...`, the recommended upgrade path will also be a custom channel/pipes, depending on complexity.

**Comment by LasseRafn** [comment]
Cool!

**Comment by taylorotwell** [comment]
Re-written / re-imagined here:

https://github.com/laravel/framework/pull/22635
